File: cpp_pt/ray.cpp
#include <cpp_pt/ray.hpp>

ray::ray(const vec& origin, const vec& direction) noexcept : ori(origin), dir(direction) {}

auto ray::origin() const noexcept -> const vec&
{
   return ori;
}
auto ray::direction() const noexcept -> const vec&
{
   return dir;
}

auto ray::position_along(double t) const noexcept -> vec
{
   return ori + dir * t;
}

File: go/tracer/ray.go
package tracer 

import "go_pt/maths"

type Ray struct {
   Origin maths.Vec3
   Direction maths.Vec3
}

func (ray *Ray) PositionAlong(t float64) maths.Vec3 {
   timeVector := maths.MultScalar(&ray.Direction, t)

   return maths.Add(&ray.Direction, &timeVector);
}
